src/storage/txn/commands/txn_heart_beat.rs

```diff
@@ -47,11 +47,63 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for TxnHeartBeat {
             !self.ctx.get_not_fill_cache(),
             context.concurrency_manager,
         );
-        let lock_ttl = txn.txn_heart_beat(self.primary_key, self.advise_ttl)?;
+        fail_point!("txn_heart_beat", |err| Err(
+            crate::storage::mvcc::Error::from(crate::storage::mvcc::txn::make_txn_error(
+                err,
+                &self.primary_key,
+                self.start_ts,
+            ))
+            .into()
+        ));
+
+        let lock_ttl: MvccResult<_> =
+            if let Some(mut lock) = txn.reader.load_lock(&self.primary_key)? {
+                if lock.ts == self.start_ts {
+                    if lock.ttl < self.advise_ttl {
+                        lock.ttl = self.advise_ttl;
+                        txn.put_lock(self.primary_key.clone(), &lock);
+                    } else {
+                        debug!(
+                            "txn_heart_beat with advise_ttl not large than current ttl";
+                            "primary_key" => %self.primary_key,
+                            "start_ts" => self.start_ts,
+                            "advise_ttl" => self.advise_ttl,
+                            "current_ttl" => lock.ttl,
+                        );
+                    }
+                    Ok(lock.ttl)
+                } else {
+                    debug!(
+                        "txn_heart_beat invoked but lock is absent";
+                        "primary_key" => %self.primary_key,
+                        "start_ts" => self.start_ts,
+                        "advise_ttl" => self.advise_ttl,
+                    );
+                    Err(MvccErrorInner::TxnLockNotFound {
+                        start_ts: self.start_ts,
+                        commit_ts: TimeStamp::zero(),
+                        key: self.primary_key.clone().into_raw()?,
+                    }
+                    .into())
+                }
+            } else {
+                debug!(
+                "txn_heart_beat invoked but lock is absent";
+                "primary_key" => %self.primary_key,
+                "start_ts" => self.start_ts,
+                "advise_ttl" => self.advise_ttl,
+                );
+                Err(MvccErrorInner::TxnLockNotFound {
+                    start_ts: self.start_ts,
+                    commit_ts: TimeStamp::zero(),
+                    key: self.primary_key.clone().into_raw()?,
+                }
+                .into())
+            };
```

Comment:
How about still put these txn logic in a separated function, and invoking it here? Maybe the control flow will be simpler

Reply:
> How about still put these txn logic in a separated function, and invoking it here? Maybe the control flow will be simpler and less change will be made on these functions

That's my origin idea, but @nrc [prefers to inline them](https://github.com/tikv/tikv/pull/8435#discussion_r469812685).

As for me, I found that there are few difference for these 4 functions, but if I want to inline some txn related functions which is used more often(and used by each other) such as `commit`/`cleanup`, things will become a little nasty.

@nrc PTAL

Reply:
I think it is nice to inline everything and then try and factor some things out. Without the inlining we might miss some good abstractions. It may be that we end up factoring the same things out as before, but I hope we can do better.

To be clear, I think we should land this PR as is and then try to do more refactoring in further PRs